### PR TITLE
[ENG-61] Fix sorting on serializer fields with source attributes

### PR DESCRIPTION
**Pull request by corbinSanders**
## Purpose
Currently, all sorting must be done on the model name. This is not intuitive to a user as some fields such as date_created have a source of created.

## Changes
api/base/filters.py - Modifying OSFOrderingFilter to support sorting on serializer fields.
api_tests/base/test_filters.py - Adding tests to ensure ordering is working correctly

## QA Notes
Test ordering on models with ?sort=date_created. The order should match ?sort=created

## Documentation
There is currently no documentation on parameterized sorting in the documentation. Perhaps there should be?

## Side Effects
Sorting by source name was not deprecated, so side effects should be minimal

## Ticket
https://openscience.atlassian.net/browse/ENG-61

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ python:
     - "2.7"
 
 sudo: false
+dist: trusty
```
Comment (UdayVarkhedkar) [inline .travis.yml]: While I understand the desire to make this change, I think it's out of scope of this Pull Request and a separate PR should be made for it or we should wait until Travis issues are resolved.

File: api/base/filters.py [hunk 2]
```diff
@@ -51,7 +51,10 @@ class OSFOrderingFilter(OrderingFilter):
     # override
     def filter_queryset(self, request, queryset, view):
         ordering = self.get_ordering(request, queryset, view)
-        if isinstance(queryset, DjangoQuerySet):
+        field_order = self.check_serializer_fields(queryset, request.query_params, view, ordering)
+        if field_order:
+            ordering = tuple(field_order)
+        if isinstance(queryset, DjangoQuerySet) and not field_order:
```
Comment (UdayVarkhedkar) [inline api/base/filters.py]:
```suggestion
       elif isinstance(queryset, DjangoQuerySet):
```

File: api/base/filters.py [hunk 3]
```diff
@@ -51,7 +51,10 @@ class OSFOrderingFilter(OrderingFilter):
     # override
     def filter_queryset(self, request, queryset, view):
         ordering = self.get_ordering(request, queryset, view)
-        if isinstance(queryset, DjangoQuerySet):
+        field_order = self.check_serializer_fields(queryset, request.query_params, view, ordering)
```
Comment (UdayVarkhedkar) [inline api/base/filters.py]: Is `request.query_params` passed and later used as opposed to using `ordering` because `ordering` has some type of field validation (https://github.com/encode/django-rest-framework/blob/master/rest_framework/filters.py#L193)?

File: api/base/filters.py [hunk 4]
```diff
@@ -66,6 +69,33 @@ def filter_queryset(self, request, queryset, view):
             return queryset.sort(*ordering)
         return queryset
 
+    def check_serializer_fields(self, queryset, query_params, view, ordering):
+        serializer_class = view.serializer_class
+        if not queryset:
+            return []
+        sorting_params = []
+        for i, field in enumerate(query_params.getlist(self.ordering_param)):
```
Comment (UdayVarkhedkar) [inline api/base/filters.py]:
```suggestion
        for field in query_params.getlist(self.ordering_param):
```
`enumerate` doesn't seem necessary here because the index provided isn't used.

File: api/base/filters.py [hunk 5]
```diff
@@ -66,6 +69,33 @@ def filter_queryset(self, request, queryset, view):
...
+            preserve_order = ''
+            # If the field is already a source, it will be returned in the ordering list
+            if field in ordering:
+                sorting_params.append(field)
+                continue
+            # The field may have a '-' before the name for descending order. This needs to be preserved
+            if field[0] == '-':
+                preserve_order = '-'
+                field = field[1:]
```
Comment (Johnetordoff) [inline api/base/filters.py]: This is nit picky, but python has a builtin func for this `.lstrip('-')`, also `preserve_order` isn't really named well might want to just make that a bool.

File: api/base/filters.py [hunk 6]
```diff
@@ -66,6 +69,33 @@ def filter_queryset(self, request, queryset, view):
...
+            if field in serializer_class._declared_fields:
+                # Checking if the field can be sorted on
+                source_field = serializer_class._declared_fields[field]
+                source_field_name = source_field.source
+                # Checking if the original sort param could be used to sort
+                if getattr(queryset[0], field, 'None') != 'None':
```
Comment (Johnetordoff) [inline api/base/filters.py]: Is this tested for fields that actually equal the string None? I'm confused as to why you've stringified this. I think you just want a try/catch here.

Comment (UdayVarkhedkar) [reply api/base/filters.py]: or
```suggestion
if getattr(queryset[0], field, None):
```

File: .travis.yml (hunk 1 above)
Comment (pattisdr) [reply .travis.yml]: drive by: this change already exists on develop now, so can be removed here -

File: api/base/filters.py [hunk 7]
```diff
@@ -51,7 +51,10 @@ class OSFOrderingFilter(OrderingFilter):
     # override
     def filter_queryset(self, request, queryset, view):
         ordering = self.get_ordering(request, queryset, view)
-        if isinstance(queryset, DjangoQuerySet):
+        field_order = self.check_serializer_fields(queryset, request.query_params, view, ordering)
+        if field_order:
+            ordering = tuple(field_order)
+        elif isinstance(queryset, DjangoQuerySet):
```
Comment (UdayVarkhedkar) [inline api/base/filters.py]: Based off of failed tests in Travis, it looks like there may be a spaces/tabs mixup here

Comment (corbinSanders) [reply api/base/filters.py]: Ran fine on my machine. Wondering if its a travis cache issue

Comment (UdayVarkhedkar) [reply api/base/filters.py]: Mind force pushing your local repo to get it to run again? Unless there's a better way to handle restarting travis' checks. I don't see a restart checks option I've seen other times I've used travis.

File: api_tests/base/test_filters.py [hunk 8]
```diff
@@ -467,6 +474,56 @@ def test_filter_queryset_handles_multiple_fields(self):
...
+        res_created = self.app.get('/{}nodes/?sort=created'.format(API_BASE), auth=user.auth)
+        res_date_created = self.app.get('/{}nodes/?sort=date_created'.format(API_BASE), auth=user.auth)
+        assert res_created.status_code == 200
+        assert res_created.json['data'] == res_date_created.json['data']
+        assert res_created.json['data'][0]['id'] == str(node_one._id)
+        assert res_date_created.json['data'][0]['id'] == node_one._id
+        assert res_created.json['data'][1]['id'] == node_two._id
+        assert res_date_created.json['data'][1]['id'] == node_two._id
+
+        # Testing both are capable of using the inverse sort sign '-'
```
Comment (UdayVarkhedkar) [inline api_tests/base/test_filters.py]: Nitpicky but I think this should simply be another test method given that it's testing something distinct

File: api/base/filters.py (hunk 7 above)
Comment (pattisdr) [reply api/base/filters.py]: i'll clear the travis cache for this PR -

**Comment by UdayVarkhedkar** [comment]
For posterity - I talked to Corbin about some small changes to tests that would help streamline them. Outside of those, everything looks good to me!

File: api/base/filters.py [hunk 9]
```diff
@@ -66,6 +69,33 @@ def filter_queryset(self, request, queryset, view):
             return queryset.sort(*ordering)
         return queryset
 
+    def check_serializer_fields(self, queryset, query_params, view, ordering):
```
Comment (pattisdr) [inline api/base/filters.py]: In general, where we're overriding DRF, I think we should try to stick as close to DRF as we can, rather than having the default be to roll our own solution. It's a well-tested framework, so I'd dig into their source code to see what we can use ourselves or what we might be able to override. Also, too many deviations are going to make this difficult to upgrade.

File: api/base/filters.py [hunk 10]
```diff
@@ -66,6 +69,33 @@ def filter_queryset(self, request, queryset, view):
...
+        serializer_class = view.serializer_class
```
Comment (pattisdr) [inline api/base/filters.py]: I'd get the `serializer_class `differently because it's not always hardcoded. 

 I'm looking at `OrderingFilter.get_default_valid_fields` which is called as part of `self.get_ordering(request, queryset, view)`, above:
```python
 if hasattr(view, 'get_serializer_class'):
            try:
                serializer_class = view.get_serializer_class()
            except AssertionError:
                # Raised by the default implementation if
                # no serializer_class was found
                serializer_class = None
        else:
            serializer_class = getattr(view, 'serializer_class', None)
```

Not all views have `serializer_class` hardcoded. Some, the serializer_class is determined based on the request method in `view.get_serializer_class()`

File: api/base/filters.py [hunk 11]
```diff
@@ -66,6 +69,33 @@ def filter_queryset(self, request, queryset, view):
...
+                sorting_params.append(field)
+                continue
+            # The field may have a '-' before the name for descending order. This needs to be preserved
+            if field[0] == '-':
+                ordering_sign = '-'
+                field = field.lstrip('-')
+            if field in serializer_class._declared_fields:
+                # Checking if the field can be sorted on
+                source_field = serializer_class._declared_fields[field]
+                source_field_name = source_field.source
```
Comment (pattisdr) [inline api/base/filters.py]: Another insight from inspecting DRF's `OrderingFilter.get_default_valid_fields`, a good reminder that the source may not be one level, but could be multiple levels deep - they are doing `field.source.replace('.', '__') `
```python
return [
            (field.source.replace('.', '__') or field_name, field.label)
            for field_name, field in serializer_class(context=context).fields.items()
            if not getattr(field, 'write_only', False) and not field.source == '*'
        ]
```

Also, they're excluding write_only fields and preventing the source from being '*'.

File: api_tests/base/test_filters.py [hunk 12]
```diff
@@ -467,6 +474,52 @@ def test_filter_queryset_handles_multiple_fields(self):
...
+        res_created = self.app.get('/{}nodes/?sort=created'.format(API_BASE), auth=user.auth)
+        res_date_created = self.app.get('/{}nodes/?sort=date_created'.format(API_BASE), auth=user.auth)
+        assert res_created.status_code == 200
+        assert res_created.json['data'] == res_date_created.json['data']
+        assert res_created.json['data'][0]['id'] == res_date_created.json['data'][0]['id']
+        assert res_created.json['data'][1]['id'] == res_date_created.json['data'][1]['id']
+
+        # Testing both are capable of using the inverse sort sign '-'
+        res_created = self.app.get('/{}nodes/?sort=-created'.format(API_BASE), auth=user.auth)
+        res_date_created = self.app.get('/{}nodes/?sort=-date_created'.format(API_BASE), auth=user.auth)
```
Comment (pattisdr) [inline api_tests/base/test_filters.py]: I'd DRY up the base nodes URL either with a method where you pass in the sort field, and returns the entire URL, or adding the sort param to the base URL (+)

File: api_tests/base/test_filters.py [hunk 13]
```diff
@@ -467,6 +474,52 @@ def test_filter_queryset_handles_multiple_fields(self):
...
+        assert res_created.status_code == 200
+        assert res_created.json['data'] == res_date_created.json['data']
+        assert res_created.json['data'][1]['id'] == res_date_created.json['data'][1]['id']
+        assert res_created.json['data'][0]['id'] == res_date_created.json['data'][0]['id']
```
Comment (pattisdr) [inline api_tests/base/test_filters.py]: ❤️ tests

File: api/base/filters.py [hunk 14]
```diff
@@ -66,6 +66,48 @@ def filter_queryset(self, request, queryset, view):
...
+        field_to_source_mapping = {}
+
+        if hasattr(view, 'get_serializer_class'):
+            serializer_class = view.get_serializer_class()
+        else:
+            serializer_class = getattr(view, 'serializer_class', None)
+
+        for field_name, field in serializer_class(context={'request': request}).fields.items():
+            if not getattr(field, 'write_only', False) and not field.source == '*' and field_name != field.source:
+                field_to_source_mapping[field_name] = field.source.replace('.', '_')
```
Comment (pattisdr) [inline api/base/filters.py]: Can you leave a comment that this doesn't handle ordering on related fields?

**Comment by pattisdr** [comment]
LGTM 🚢

**Comment by pattisdr** [comment]
merging into feature/staging-1 for testing
